File: java/ValidateBinarySearchTree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */


public class Solution {
    
    private class resultType {
        boolean isBST;
        int min;
        int max;
        public resultType(boolean isBST, int min, int max) {
            this.isBST = isBST;
            this.min = min;
            this.max = max;
        }
    }
    
    /*
     * @param root: The root of binary tree.
     * @return: True if the binary tree is BST, or false
     */
    public boolean isValidBST(TreeNode root) {
        // write your code here
        if (root == null) {
            return true;
        }

        return helper(root).isBST;
    }
    
    private resultType helper(TreeNode node){
        if (node == null) {
            return new resultType(true, Integer.MAX_VALUE, Integer.MIN_VALUE);
        }
        
        resultType left = helper(node.left);
        resultType right = helper(node.right);
        
        if (!left.isBST || !right.isBST) {
            return new resultType(false, 0, 0);
        }
        
        if (node.left != null && left.max >= node.val ||
            node.right != null && right.min <= node.val) {
            return new resultType(false, 0, 0);
        }
        
        return new resultType(true, Math.min(left.min,node.val), 
                                    Math.max(right.max,node.val));
    }
}

File: python/validateBinarySearchTree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
import sys

class resultType:
    def __init__(self, isBst, min, max):
        self.isBst = isBst
        self.min = min
        self.max = max

class Solution:
    """
    @param: root: The root of binary tree.
    @return: True if the binary tree is BST, or false
    """
    def isValidBST(self, root):
        # write your code here
        return self.helper(root).isBst


    def helper(self, root):
        if root is None:
            return resultType(True, sys.maxsize, -sys.maxsize)

        left = self.helper(root.left)
        right = self.helper(root.right)

        if not left.isBst or not right.isBst:
            return resultType(False, 0, 0)

        if root.left is not None and left.max >= root.val or root.right is not None and right.min <= root.val:
            return resultType(False, 0, 0)

        return resultType(True, min(left.min, root.val), max(right.max, root.val))